[Quant] [Inductor] Enable the Inductor Lowering of QConv2d post op hardtanh

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114547
* __->__ #114580
* #114579
* #114578

**Summary**
Enable the fusion pattern of `QConv2d -> hardtanh` lowering to `hardtanh` as `QConv2d` post operator. 

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_relu6_cpu
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_hardtanh_cpu

python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_relu6
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_hardtanh
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

leslie-fang-intel committed:
Enable the Inductor Lowering of QConv2d post op hardtanh

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -521,6 +532,9 @@ def __init__(self, op_name: str, scalars_attr=None, algorithm_attr=None):
             UnaryAttr("relu", [], ""): generate_pattern_with_unary(
                 dequantize_qconv_pt2e_pattern, aten.relu.default
             ),
+            UnaryAttr("hardtanh", [], ""): _hardtanh_fusion(
+                dequantize_qconv_pt2e_pattern
+            ),
```
Can we extend `generate_pattern_with_unary` to support "hardtanh" fusion rather than relying on a pattern specific function?

leslie-fang-intel committed:
Update on "[Quant] [Inductor] Enable the Inductor Lowering of QConv2d post op hardtanh"


**Summary**
Enable the fusion pattern of `QConv2d -> hardtanh` lowering to `hardtanh` as `QConv2d` post operator. 

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_relu6_cpu
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_hardtanh_cpu

python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_relu6
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_hardtanh
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [Inductor] Enable the Inductor Lowering of QConv2d post op hardtanh"


**Summary**
Enable the fusion pattern of `QConv2d -> hardtanh` lowering to `hardtanh` as `QConv2d` post operator. 

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_relu6_cpu
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_hardtanh_cpu

python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_relu6
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_hardtanh
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [Inductor] Enable the Inductor Lowering of QConv2d post op hardtanh"


**Summary**
Enable the fusion pattern of `QConv2d -> hardtanh` lowering to `hardtanh` as `QConv2d` post operator. 

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_relu6_cpu
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_hardtanh_cpu

python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_relu6
python -m pytest test_mkldnn_pattern_matcher.py -k test_qat_qconv2d_hardtanh
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
@pytorchbot merge
